x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/heuristic/PValueScore.java

```diff
@@ -0,0 +1,187 @@
+        out.writeBoolean(backgroundIsSuperset);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if ((obj instanceof PValueScore) == false) {
+            return false;
+        }
+        return super.equals(obj);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = NAME.hashCode();
+        result = 31 * result + super.hashCode();
+        return result;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(BACKGROUND_IS_SUPERSET.getPreferredName(), backgroundIsSuperset);
+        builder.endObject();
+        return builder;
+    }
+
+    public static SignificanceHeuristic parse(XContentParser parser) throws IOException {
+        return PARSER.apply(parser, null);
+    }
+
+    /**
+     *  This finds the p-value that the frequency of a category is unchanged on set subset assuming
+     *  we observe subsetFreq out of subset values in total relative to set supersetFreq where it accounts
+     *  supersetFreq out of supersetSize total.
+     *
+     *  This assumes that each sample is an independent Bernoulli trial and computes the p-value
+     *  under the null hypothesis that the probabilities are the same. Note that the independence
+     *  assumption is quite strong and can lead to low p-values even if the fractions are very
+     *  similar if there are many trials. We arrange for small differences in frequency to always
+     *  have large p-values. We also artificially increase the p-value of when the probability
+     *  of the category is very small.
+     *
+     *  NOTE: Since in the original calculation of `p-value`, smaller indicates more significance, the value actual value returned
+     *        is `log(-p-value)`. To get the original p-value from the score, simply calculate `exp(-retval)`
+     *
+     * @return log(-p-value)
+     */
+    @Override
+    public double getScore(long subsetFreq, long subsetSize, long supersetFreq, long supersetSize) {
+        Frequencies frequencies = computeNxys(subsetFreq, subsetSize, supersetFreq, supersetSize, "PValueScore");
+        double docsContainTermInClass = frequencies.N11;
+        double allDocsInClass = frequencies.N_1;
+        double docsContainTermNotInClass = frequencies.N10;
+        double allDocsNotInClass = frequencies.N_0;
+
+        if (docsContainTermInClass * allDocsNotInClass <= allDocsInClass * docsContainTermNotInClass) {
+            return 0.0;
+        }
+
+        if (allDocsNotInClass == 0L || allDocsInClass == 0L) {
+            return 0.0;
+        }
+
+        // Adjust counts to ignore ratio changes which are less than 5%
+        // casting to `long` to round down to nearest whole number
+        if (docsContainTermInClass * allDocsNotInClass < docsContainTermNotInClass * allDocsInClass) {
```

Comment:
Isn't this term the same as line 108 with the order switched on the RHS

`allDocsInClass * docsContainTermNotInClass` 
`docsContainTermNotInClass * allDocsInClass`

Reply:
It is see [this](https://github.com/elastic/elasticsearch/pull/75313#discussion_r669984480) comment. This was lifted  from something which could test for both increases and decreases in frequency, but here we only care about increases in frequency so this simplifies.

Reply:
latest commit addresses this.